zerver/lib/test_classes.py

```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
 
     def setUpBeforeMigration(self, apps: StateApps) -> None:
         pass  # nocoverage
+
+class LDAPTestCase(ZulipTestCase):
+    """
+    Base class for LDAP tests. Takes care of the mock_ldap setup, loads
+    a directory from fixtures/ldap/directory.json with various entries to be
+    used by tests. If a test wants to specify its own directory, it can just replace
+    self.mock_ldap.directory with its own content, but in most cases it should be
+    enough to use change_user_attr to make simple modifications to the pre-loaded
+    directory. If new user entries are needed to test for some additional unusual
+    scenario, it's most likely best to add that to directory.json.
+    """
+
+    def setUp(self) -> None:
+        directory = ujson.loads(self.fixture_data("directory.json", type="ldap"))
+
+        # Load binary attributes. If in "directory", an attribute as its value
+        # has a string starting with "file:", the rest of the string is assumed
+        # to be a path to the file from which binary data should be loaded,
+        # as the actual value of the attribute in ldap.
+        for dn, attrs in directory.items():
+            for attr, value in attrs.items():
+                if isinstance(value, str) and value.startswith("file:"):
+                    with open(value[5:], 'rb') as f:
+                        attrs[attr] = [f.read(), ]
+
+        ldap_patcher = mock.patch('django_auth_ldap.config.ldap.initialize')
+        self.mock_initialize = ldap_patcher.start()
+        self.mock_ldap = MockLDAP(directory)
+        self.mock_initialize.return_value = self.mock_ldap
+
+    def tearDown(self) -> None:
+        self.mock_ldap.reset()
+        self.mock_initialize.stop()
+
+    def settings(self, **kwargs: Any) -> Any:
+        """
+        Extend self.settings for LDAP tests to set various LDAP defaults if not explicitly specified:
+        The server URI, and how to find users by username or email.
+        """
+        if "AUTH_LDAP_USER_SEARCH" not in kwargs:
+            kwargs["AUTH_LDAP_USER_SEARCH"] = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                         ldap.SCOPE_ONELEVEL, "(uid=%(user)s)")
```

Comment:
I think this patching process is somewhat dangerous, because it affects all `with self.settings()` calls in the test case, not just those that mean to be affecting LDAP settings.  I think it might be cleaner to just use a special value, e.g. `AUTH_LDAP_USER_SEARCH=self.default_ldap_search_config()`, rather than having this hack.

Reply:
(I think this lets us remove this overridden function entirely)

Reply:
Instead of overriding the settings() method like this, I figured that it seems most sensible to just put the configuration in ``test_settings.py``, the straight-forward way. I hope that's a more reasonable approach - I don't think having those settings in all tests affects anything important - and the tests that don't want them, or want different ones, should override them explicitly in my opinion.